File: 2015/day4.jl
using MD5

function findcoin(key::String, targetprefix::String)
    n = 1

    while true
        inputstring = key * string(n)
        hash = md5(inputstring)
        hashhex = bytes2hex(hash)

        if startswith(hashhex, targetprefix)
            println("Found a match for '$(key)': $(hashhex)")
            return n
        end

        n += 1
    end
end

function findcoin_threaded(key::String, targetprefix::String)
    println("Running search on $(Threads.nthreads()) threads...")
    chunksize = 10_000
    n = 1

    while true
        result = Threads.Atomic{Int}(0)  # Thread safe integer

        Threads.@threads for i in n:(n + chunksize - 1)  # Split range into chunks
            if result[] > 0  # Check if another thread found the result
                continue
            end

            inputstring = key * string(i)
            hash = md5(inputstring)
            hashhex = bytes2hex(hash)

            if startswith(hashhex, targetprefix)
                Threads.atomic_cas!(result, 0, i)  # compare-and-swap
            end
        end

        if result[] > 0
            println("Found a match for '$(key)': $(bytes2hex(md5(key * string(result[]))))")
            return result[]
        end

        n += chunksize
    end
end

function main()
    givenkey = read("2015/inputs/day4.txt", String)

    println("Single threaded search:")
    time1 = @elapsed lowestnumber = findcoin(givenkey, "00000")
    println("The lowest positive number to match the key is: '$(lowestnumber)' (found in $(round(time1, digits=3))s)")
    time2 = @elapsed lowestnumber_six = findcoin(givenkey, "000000")
    println("The lowest positive number to match the key with 6 zeros is: '$(lowestnumber_six)' (found in $(round(time2, digits=3))s)\n")

    println("Multi threaded search:")
    time3 = @elapsed lowestnumber_threaded = findcoin_threaded(givenkey, "00000")
    println("The lowest positive number to match the key is: '$(lowestnumber_threaded)' (found in $(round(time3, digits=3))s)")
    time4 = @elapsed lowestnumber_threaded_six = findcoin_threaded(givenkey, "000000")
    println("The lowest positive number to match the key with 6 zeros is: '$(lowestnumber_threaded_six)' (found in $(round(time4, digits=3))s)")
end

main()

File: 2015/day4.m
function day4()
    givenKey = fileread('inputs/day4.txt');

    fprintf("Executing single threaded search...\n");
    tic;
    lowestNumber1 = findCoin(givenKey, '00000');
    time1 = toc;
    fprintf("The lowest positive integer to match the given key (5 zero start): '%d' (found in %.3fs)\n", lowestNumber1, time1);

    tic;
    lowestNumber2 = findCoin(givenKey, '000000');
    time2 = toc;
    fprintf("The lowest positive integer to match the given key (6 zero start): '%d' (found in %.3fs)\n", lowestNumber2, time2);

    fprintf("Executing multi threaded search...\n");
    tic;
    lowestNumber3 = findCoinThreaded(givenKey, '00000');
    time3 = toc;
    fprintf("The lowest positive integer to match the given key (5 zero start): '%d' (found in %.3fs)\n", lowestNumber3, time3);

    tic;
    lowestNumber4 = findCoinThreaded(givenKey, '000000');
    time4 = toc;
    fprintf("The lowest positive integer to match the given key (6 zero start): '%d' (found in %.3fs)\n", lowestNumber4, time4);
end


function n = findCoin(key, targetPrefix)
    n = 1;

    while true
        inputString = [key, num2str(n)];
        hashHex = mlreportgen.utils.hash(inputString);

        if startsWith(hashHex, targetPrefix)
            fprintf("Found a match for '%s': %s\n", key, hashHex);
            return;
        end

        n = n + 1;
    end
end

function n = findCoinThreaded(key, targetPrefix)
    n = 1;
    chunkSize = 10000;

    while true
        results = zeros(chunkSize, 1);

        % parallel for loop
        parfor i = 1:chunkSize
            nCurrent = n + i - 1;
            inputString = [key, num2str(nCurrent)];
            hashHex = mlreportgen.utils.hash(inputString);

            if startsWith(hashHex, targetPrefix)
                results(i) = nCurrent;
            end
        end

        % check if a worker found a match
        matches = results(results > 0);
        if ~isempty(matches)
            result = min(matches);
            foundHash = mlreportgen.utils.hash([key, num2str(result)]);
            fprintf("Found a match for '%s': %s\n", key, foundHash);
            n = result;
            return;
        end

        n = n + chunkSize;
    end
end